### ISSUE TITLE
[Bug] Search engine is placed at the bottom of the list and set as default when added from the "Add search engine"

### ISSUE THREAD

**Issue by ebalazs-sv** [mozilla-mobile/fenix#17044]
### Precondition
Have Google set as default search engine
## Steps to reproduce
1. Launch Fenix and go to Settings -> Search;
2. Delete "Google" from the Default search engine list;
3. Tap "Add search engine" and add "Google";
4. Observe where "Google" is placed in the list;
### Expected behavior
"Google" is placed at the top of the list, where it was before was deleted. Also it is not set as default.
### Actual behavior
"Google" is placed at the bottom of the list and is set as default.

### Notes
After a restart, "Google" is placed at the top but is still set as default.
This issue is reproducible with other search engines from the list.
Video:
![20201215_130532](https://user-images.githubusercontent.com/48799384/102207206-65656780-3ed6-11eb-8138-2d76dc53c63e.gif)

### Device information

* Android device: Huawei P9 Lite (Android 7), OnePlus 5T (android 9);
* Fenix version: Nightly 201215 05:02 (Build #2015781481) GV 85 from 12/15
Not reproducible on the latest Beta and Release.

**Comment by ebalazs-sv** [comment]
This issue is also reproducible on 85.0.0-beta.3 from 12/21 with Samsung Galaxy Tab S3 (Android 8).

**Comment by softvision-miralobontiu** [comment]
This issue is reproducible on Beta 85.0.0-beta.6 with Sony Xperia Z5 Premium (Android 7.1.1).

**Comment by mcarare** [comment]
Note that this happens with any default search engine that is deleted, not just Google.

**Comment by mcarare** [comment]
As for the engine list order, I am not aware of a rule that says Google should be on top. Currently, they are ordered by type (default from region and additional default - the ones in the additional list), and by the time they were added or readded to the list. If UX does not mention any other order (alphabetical, etc) I do not consider it to be a bug the fact that it is not placed at the top of the list.

**Comment by abodea** [comment]
Removing `eng:qa:needed` until further notice due to the reverts from above.

**Comment by eliserichards** [comment]
We should double-check with the Search team with regards to the order of the search engines when deleted/re-added (for example, if we delete google and re-add it, should it still be at the top?)

**Comment by mcarare** [comment]
@pocmo Do you have more insights on what the desired behavior would be on deleting the default engine? If there is no decision to be made regarding an automatic selection of another engine, maybe we can prompt the user to either:

1. Cannot delete default engine and should choose another one before deleting the current default.
2. Provide the engine list in the dialog and select one before continuing with the deletion.

Ux should also weigh in on this.

**Comment by amedyne** [comment]
@topotropic, @brampitoyo any thoughts on this?

**Comment by pocmo** [comment]
> @pocmo Do you have more insights on what the desired behavior would be on deleting the default engine? If there is no decision to be made regarding an automatic selection of another engine, maybe we can prompt the user to either:

There is a fallback decision being made by the search code in AC, it's just not *persisted* as user choice, so that it can still change. See how "Bing" gets selected in the GIF in the first comment.

I understand this issue here only as a "order" issue. What we do not do currently is remember the original order, so once we add "Google" back it will be at the end of the list instead of at the top. This is only a temporary issue though, once the list is reloaded from disk it will again be in the correct order again.

So overall I think this is only a minor problem since it is only a visual "glitch" in this specific scenario (removing the search engine and immediately adding it back) that resolves itself. What we could do though, is to save the order we got from the search config in the state too and use that to always order the list. 🤔

**Comment by Mugurell** [comment]
>  What we could do though, is to save the order we got from the search config in the state too and use that to always order the list.

If @brampitoyo confirms maybe with this we can put a closure on the ticket.

**Comment by pocmo** [comment]
I drafted this here: https://github.com/mozilla-mobile/android-components/pull/9735

**Comment by brampitoyo** [comment]
@eliserichards wrote:
> with regards to the order of the search engines when deleted/re-added (for example, if we delete google and re-add it, should it still be at the top?

This is desktop’s behaviour that we should also follow.

- When a default search engine is removed, the next search engine on the search config order should be set as default
    - Example: My default is `Google`
    - I delete it
    - Result: Set `Bing` as default
- When a search engine is re-added, it should appear in the original search config order, then be set as the new default
    - Example: My default is `DuckDuckGo`
    - I delete it
    - I re-add it
    - Result: `DuckDuckGo` should reappear above `Wikipedia` and below `Amazon.com`. Set `DuckDuckGo` as default.

-----

Additionally, we don’t have a `Restore Default Search Engine` functionality, but may want to consider it in the future. This is how it should work.

- When `Restore Default Search Engine` is executed, the original search config order should be restored. The default should also be restored.
    - Example: My default is `Wikipedia`
    - I delete it and other search engines
    - I tap on `Restore Default Search Engine`
    - Result: original search config order is restored (Google, Bing, Amazon.com…). Set `Google` as default.

**Comment by mcarare** [comment]
Talked with @pocmo offline. This PR in AC should handle this issue: https://github.com/mozilla-mobile/android-components/pull/9735.

**Comment by abodea** [comment]
Verified as fixed on the latest Nightly 2/25 with Google Pixel 4 XL (Android 11).
Based on the comments above I did some cases:

1. Google
2. Bing
3. Amazon
4. DuckDuckGo
5. Wikipedia
 
- Amazon as default -> deleted Bing and Amazon - > Google was set by default - > Restored Bing and Amazon - > Amazon and Bing were placed in their initial place and Amazon was set as default.
- Bing as default -> deleted Google and Bing - > Amazon was set as default - > Restored Bing and Google - > Google and Bing were placed in their initial place and Bing was set as default.
I did other scenarios with all the search engines but they work as expected.
